conda_lock/conda_solver.py

```diff
@@ -0,0 +1,490 @@
+    planned = {
+        action["name"]: LockedDependency(
+            name=action["name"],
+            version=action["version"],
+            manager="conda",
+            platform=platform,
+            dependencies={
+                item.split()[0]: " ".join(item.split(" ")[1:])
+                for item in action.get("depends") or []
+            },
+            url=action["url"],
+            # NB: virtual packages may have no hash
+            hash=action.get("md5", ""),
+        )
+        for action in dry_run_install["actions"]["FETCH"]
+    }
+
+    # propagate categories from explicit to transitive dependencies
+    _apply_categories({k: v for k, v in specs.items() if v.manager == "conda"}, planned)
+
+    return planned
+
+
+def _reconstruct_fetch_actions(
+    conda: PathLike, platform: str, dry_run_install: DryRunInstall
+) -> DryRunInstall:
+    """
+    Conda may choose to link a previously downloaded distribution from pkgs_dirs rather
+    than downloading a fresh one. Find the repodata record in existing distributions
+    that have only a LINK action, and use it to synthesize an equivalent FETCH action.
+    """
+    if "LINK" not in dry_run_install["actions"]:
+        dry_run_install["actions"]["LINK"] = []
+    if "FETCH" not in dry_run_install["actions"]:
+        dry_run_install["actions"]["FETCH"] = []
+
+    link_actions = {p["name"]: p for p in dry_run_install["actions"]["LINK"]}
+    fetch_actions = {p["name"]: p for p in dry_run_install["actions"]["FETCH"]}
+    link_only_names = set(link_actions.keys()).difference(fetch_actions.keys())
+    # NB: micromamba does not support info --json, nor does it appear to honor pkgs_dirs from .condarc
+    if not is_micromamba(conda):
+        if link_only_names:
+            pkgs_dirs = [
+                pathlib.Path(d)
+                for d in json.loads(
+                    subprocess.check_output(
+                        [str(conda), "info", "--json"], env=conda_env_override(platform)
+                    )
+                )["pkgs_dirs"]
+            ]
+        else:
+            pkgs_dirs = []
+
+        for link_pkg_name in link_only_names:
+            link_action = link_actions[link_pkg_name]
+            for pkgs_dir in pkgs_dirs:
+                record = (
+                    pkgs_dir
+                    / link_action["dist_name"]
+                    / "info"
+                    / "repodata_record.json"
+                )
+                if record.exists():
+                    with open(record) as f:
+                        repodata: FetchAction = json.load(f)
+                    break
+            else:
+                raise FileExistsError(
+                    f'Distribution \'{link_action["dist_name"]}\' not found in pkgs_dirs {pkgs_dirs}'
+                )
+            dry_run_install["actions"]["FETCH"].append(repodata)
+    else:
+        # NB: micromamba LINK actions are secretly equivalent to FETCH actions
```

Comment:
Good discovery cc @wolfv ?

Reply:
I don't think that's true.
if there is no FETCH it just means that the package is already in the cache.

If you clean the cache (`pkgs` directory) the FETCH should appear.

Reply:
This comment was somewhat misleading. What I meant is that in mainline conda, LINK actions contain only a subset of the metadata we need to harvest for the lock file, notably missing `md5` and `depends`, so I have to go digging around `pkgs_dirs` to extract them from repodata.json. Micromamba appears to use the same structure for both LINK and FETCH, so I can just copy the contents of the LINK action into a FETCH action, and then later harvest the FETCH actions without needing to know which implementation generated them, or whether or not `pkgs_dirs` was configured.

I've tried to clarify the comment in the source.